src/RestNews/DB/ProcessRequest.hs

```diff
@@ -212,32 +214,131 @@ createCategory sessionRun connection createCategoryRequest = do
             Left sessionError -> case getErrorCode sessionError of
                 Just PSQL_STRING_DATA_RIGHT_TRUNCATION -> H . Right . Left $ encode eCategoryNameMaxBoundOverflow
-                Just PSQL_FOREIGN_KEY_VIOLATION -> H . Right . Left $ encode eParentCategoryDoesNotExist
-                Just UnexpectedAmountOfRowsOrUnexpectedNull -> H . Right . Left $ encode eNoSuchCategory
+                Just PSQL_FOREIGN_KEY_VIOLATION ->
+                    case maybe_parent_id' of
+                        Just parent_id' -> H . Right . Left . encode . makeNoSuchCategory . pack $ show parent_id'
+                        _ ->  H . Left $ show sessionError
+                Just UnexpectedAmountOfRowsOrUnexpectedNull ->
+                    H . Right . Left . encode . makeNoSuchCategory . pack $ show category_id'
                 _ -> H . Left $ show sessionError
         )
 
+
+sameCategoryIdCheck :: MonadIO m =>
+    Int32
+    -> Int32
+    -> m (Either (Either UnhandledError ErrorForUser) ())
+sameCategoryIdCheck category_id' parent_id' =
+    if category_id' == parent_id'
+        then pure . Left . Right $ encode eSameParentId
+        else pure $ Right ()
+
+
+isCategoryExist :: MonadIO m =>
+    Connection
+    -> Int32
+    -> m (Either Session.QueryError Bool)
+isCategoryExist connection category_id' =
+    liftIO
+        $ Session.run
+            (Session.statement category_id' DBR.isCategoryExist)
+            connection
+
+parentCategoryExistCheck :: MonadIO m =>
+    Connection
+    -> Int32
+    -> m (Either (Either UnhandledError ErrorForUser) ())
+parentCategoryExistCheck connection parent_id' = do
+    
+    sessionResults <- isCategoryExist connection parent_id'
+
+    pure $ case sessionResults of
+        Right isCategoryExist' ->
+            if isCategoryExist'
+                then Right ()
+                else Left . Right . encode . makeNoSuchCategory . pack $ show parent_id'
+        Left sessionError -> Left . Left $ show sessionError
+
+
+getCategoryDescendants :: MonadIO m =>
+    Connection
+    -> Int32
+    -> m (Either Session.QueryError (Vector Int32))
+getCategoryDescendants connection category_id' =
+    liftIO
+        $ Session.run
+            (Session.statement category_id' DBR.getCategoryDescendants)
+            connection
+
+
+parentIdDescendantCheck :: MonadIO m =>
+    Connection
+    -> (Int32, Int32)
+    -> m (Either (Either UnhandledError ErrorForUser) ())
+parentIdDescendantCheck connection (category_id', parent_id') = do
+
+    eitherDescendants <- getCategoryDescendants connection category_id'
+
+    pure $ case eitherDescendants of
+
+        Left sessionError -> Left . Left $ show sessionError
+
+        Right descendants ->
+            let parentIdDescendantCheck' = Data.Vector.elem parent_id' descendants
+            in if parentIdDescendantCheck'
+                then Left . Right $ encode eParentIdIsDescendant
+                else Right ()
```

Comment:
Внутри `case` можно не писать `if`, кейс умеет булевы условия:

```suggestion
        Right descendants
            | parent_id' `Data.Vector.elem` descendants -> Left . Right $ encode eParentIdIsDescendant
            | otherwise -> Right ()
```

И еще `elem` часто пишут инфиксно, но это кому как нравится.